starcluster/streaming_node_add.py

```diff
@@ -0,0 +1,191 @@
+        if instance_ids:
+            log.info("Instance ids:" + str(instance_ids))
+            for instance_id in instance_ids:
+                self.unpropagated_instances.append(
+                    UnpropagatedInstance(instance_id))
+        if self.spots:
+            self.spots = \
+                self.cluster.ec2.cancel_stuck_spot_instance_request(self.spots)
+        if self.spots:
+            log.info("Still waiting for spots: " + str(self.spots))
+
+    def stream_unpropagated_instances(self):
+        if not self.unpropagated_instances:
+            return
+
+        propagated_instance_ids = self.cluster.ec2.get_propagated_instances(
+            [s.id for s in self. unpropagated_instances])
+        self.unpropagated_instances = utils.filter_move(
+            lambda i: i.id not in propagated_instance_ids,
+            self.unpropagated_instances, self.instances)
+        if self.unpropagated_instances:
+            log.info("Still waiting for unpropagated instances: "
+                     + str(self.unpropagated_instances))
+        self.instances = self.cluster.get_nodes_or_raise(nodes=self.instances)
+
+    def stream_update_nrm(self):
+        for instance in self.instances:
+            if instance.id not in self.instances_nrm:
+                nrm_cls = partial(NodeRecoveryManager,
+                                  reboot_interval=self.reboot_interval,
+                                  n_reboot_restart=self.n_reboot_restart)
+                if isinstance(instance, Node):
+                    nrm = nrm_cls(instance)
+                else:
+                    nrm = nrm_cls(Node(instance, self.cluster.key_location))
+                self.instances_nrm[instance.id] = nrm
+
+    def stream_instances(self, ready_instances):
+        if not self.instances:
+            return
+
+        ssh_up = self.cluster.pool.map(lambda i: i.is_up(), self.instances)
+        zip_instances = utils.filter_move(
+            lambda i: i[0].state != 'running' or not i[1],
+            zip(self.instances, ssh_up), ready_instances,
+            lambda i: i[0])
+        self.instances = [i[0] for i in zip_instances]
+        if self.instances:
+            log.info("Still waiting for instances: " + str(self.instances))
+
+    def stream_manage_reboots(self):
+        dead_instances = []
+        self.instances = \
+            utils.filter_move(lambda i: self.instances_nrm[i.id].check(),
+                              self.instances, dead_instances)
+        for instance in dead_instances:
+            del self.instances_nrm[instance.id]
+
+    def stream_ready_instances(self, ready_instances):
+        for ready_instance in ready_instances:
+            log.info("Adding node: %s" % ready_instance.alias)
+            up_nodes = filter(lambda n: n.is_up(), self.cluster.nodes)
+            try:
+                self.cluster.run_plugins(method_name="on_add_node",
+                                         node=ready_instance, nodes=up_nodes)
```

Comment:
@vasisht You are correct for the core version, but the vanilla_improvements one has the "nodes" argument.
https://github.com/datacratic/StarCluster/blob/54b895b40975beaed547be024db1767dabe87971/starcluster/cluster.py#L1977